
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>csg.genome package &#8212; csglib 0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="csg-genome-package">
<h1>csg.genome package<a class="headerlink" href="#csg-genome-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-csg.genome.coordinates">
<span id="csg-genome-coordinates-module"></span><h2>csg.genome.coordinates module<a class="headerlink" href="#module-csg.genome.coordinates" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="csg.genome.coordinates.assign_locus">
<code class="descclassname">csg.genome.coordinates.</code><code class="descname">assign_locus</code><span class="sig-paren">(</span><em>df</em>, <em>col_chrom</em>, <em>col_pos</em>, <em>locus_name=’LOCUS’</em>, <em>dist=250000</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genome.coordinates.assign_locus" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns a locus number to each variant in a data frame of variants with chromosome and position specified.</p>
<p>A column will be inserted containing the locus number assigned to each variant.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>df</strong> – data frame, each row is a variant.</li>
<li><strong>chrom</strong> – chromosome column name.</li>
<li><strong>pos</strong> – position column name.</li>
<li><strong>locus_name</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) – What should we call the locus number column?</li>
<li><strong>dist</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) – base-pair distance between variants required to assign a new locus number</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">with an added column specifying the locus number</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">DataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="csg.genome.coordinates.chrom_int">
<code class="descclassname">csg.genome.coordinates.</code><code class="descname">chrom_int</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genome.coordinates.chrom_int" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a chromosome string (e.g. chr9 or chrX) to an integer</p>
<dl class="docutils">
<dt>The conversion ends up as:</dt>
<dd>X -&gt; 23
Y -&gt; 24
MT -&gt; 25
XY -&gt; 26</dd>
</dl>
<p>The chromosome string can be either chromX, chrX, or just X, for example.</p>
</dd></dl>

<dl class="function">
<dt id="csg.genome.coordinates.parse_bp">
<code class="descclassname">csg.genome.coordinates.</code><code class="descname">parse_bp</code><span class="sig-paren">(</span><em>fstr</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genome.coordinates.parse_bp" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a unit of genomic length from a string. Possible strings:</p>
<p>1.25 MB or 1.25 mb = 1,250,000 bp
300 kb = 300,000 bp
20000 = 20,000 bp</p>
</dd></dl>

</div>
<div class="section" id="csg-genome-intervals-module">
<h2>csg.genome.intervals module<a class="headerlink" href="#csg-genome-intervals-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-csg.genome">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-csg.genome" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">csg.genome package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-csg.genome.coordinates">csg.genome.coordinates module</a></li>
<li><a class="reference internal" href="#csg-genome-intervals-module">csg.genome.intervals module</a></li>
<li><a class="reference internal" href="#module-csg.genome">Module contents</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Ryan Welch.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>