
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>csg.genetics.ld package &#8212; csglib 0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="csg-genetics-ld-package">
<h1>csg.genetics.ld package<a class="headerlink" href="#csg-genetics-ld-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-csg.genetics.ld.pyld">
<span id="csg-genetics-ld-pyld-module"></span><h2>csg.genetics.ld.pyld module<a class="headerlink" href="#module-csg.genetics.ld.pyld" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="csg.genetics.ld.pyld.Haplotypes">
<em class="property">class </em><code class="descclassname">csg.genetics.ld.pyld.</code><code class="descname">Haplotypes</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes"><code class="xref py py-class docutils literal"><span class="pre">csg.genetics.ld.pyld.Haplotypes</span></code></a></p>
<p>This class represent phased genotypes at chromosomal region.</p>
<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.size">
<code class="descname">size</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.size" title="Permalink to this definition">¶</a></dt>
<dd><p><em>int</em> – number of variants (M).</p>
</dd></dl>

<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.chrom">
<code class="descname">chrom</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.chrom" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list</em> – list of length M with chromosome names.</p>
</dd></dl>

<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.position">
<code class="descname">position</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.position" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list</em> – list of length M with chromosomal positions (long type).</p>
</dd></dl>

<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.ref">
<code class="descname">ref</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.ref" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list</em> – list of length M with reference alleles (string type).</p>
</dd></dl>

<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.alt">
<code class="descname">alt</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.alt" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list</em> – list of length M with alternate alleles (string type).</p>
</dd></dl>

<dl class="attribute">
<dt id="csg.genetics.ld.pyld.Haplotypes.haplotypes">
<code class="descname">haplotypes</code><a class="headerlink" href="#csg.genetics.ld.pyld.Haplotypes.haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p><em>numpy.ndarray</em> – M x N matrix of genotypes (coded as 0 and 1), where N is number of individuals times 2.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="csg.genetics.ld.pyld.LD">
<em class="property">class </em><code class="descclassname">csg.genetics.ld.pyld.</code><code class="descname">LD</code><a class="headerlink" href="#csg.genetics.ld.pyld.LD" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="csg.genetics.ld.pyld.LD.add_vcf">
<code class="descname">add_vcf</code><span class="sig-paren">(</span><em>vcf</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.add_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens VCF file with phased genotypes.</p>
<p>Phased genotypes must be stored in VCF format: in single VCF file, or multiple VCF files split by chromosome.
Each VCF file must be indexed using tabix.
If VCF files are split by chromosome, then envoke this method for each file individually.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>vcf</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – path to VCF file which is compressed using bgzip and indexed using tabix.</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">Exception</span></code> – Raises an exception if: (a) no VCF header; (b) numbers of individuals in multiple VCF files are different.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.compute_freq">
<code class="descname">compute_freq</code><span class="sig-paren">(</span><em>haplotypes</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.compute_freq" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes alternate allele frequencies for set of variants.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>haplotypes</strong> (<a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes"><em>Haplotypes</em></a>) – genotypes of variants.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">array of size haplotypes.size with alternate allele frequencies.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">numpy.ndarray</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.compute_r_cross">
<code class="descname">compute_r_cross</code><span class="sig-paren">(</span><em>haplotypes1</em>, <em>haplotypes2</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.compute_r_cross" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes r coefficient of linkage-disequilibrium between pairs of variants in two sets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>haplotypes1</strong> (<a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes"><em>Haplotypes</em></a>) – genotypes for variants in first set.</li>
<li><strong>haplotypes2</strong> (<a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes"><em>Haplotypes</em></a>) – genotypes for variants in second set.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><dl class="docutils">
<dt>haplotypes1.size x haplotypes2.size matrix of r coefficients of linkage-disequilibrium.</dt>
<dd><p class="first last">Matrix rows correspond to variants in first set, while columns correspond to variants in second set.
Both rows and columns have the same order as variants in haplotypes1 and haplotypes2 objects, correspondingly.</p>
</dd>
</dl>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">numpy.ndarray</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.compute_r_pairwise">
<code class="descname">compute_r_pairwise</code><span class="sig-paren">(</span><em>haplotypes</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.compute_r_pairwise" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes r coefficient of linkage-disequilibrium between all variants in the given set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>haplotypes</strong> (<a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes"><em>Haplotypes</em></a>) – genotypes for a set of variants.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="docutils">
<dt>haplotypes.size x haplotypes.size symmetric matrix of r coefficients of linkage disequilibrium.</dt>
<dd>Matrix rows and columns have the same order as variants in haplotypes object.</dd>
</dl>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">numpy.ndarray</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.get_region_haplotypes">
<code class="descname">get_region_haplotypes</code><span class="sig-paren">(</span><em>chrom</em>, <em>start_position</em>, <em>end_position</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.get_region_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads phased genotypes at specified chromosomal region.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – chromosome name.</li>
<li><strong>start_position</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#long" title="(in Python v2.7)"><em>long</em></a>) – start position of chromosomal region in base pairs.</li>
<li><strong>end_position</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#long" title="(in Python v2.7)"><em>long</em></a>) – end position of chromosomal region in base pairs.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">phased genotypes at specified chromosomal region.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes">Haplotypes</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.get_variant_haplotypes">
<code class="descname">get_variant_haplotypes</code><span class="sig-paren">(</span><em>chrom</em>, <em>position</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.get_variant_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads phased genotypes at specified chromosome and position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – chromosome name.</li>
<li><strong>position</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#long" title="(in Python v2.7)"><em>long</em></a>) – chromosomal position in base pairs.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">phased genotypes at specified chromosome and position.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes">Haplotypes</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.get_variant_haplotypes_strict">
<code class="descname">get_variant_haplotypes_strict</code><span class="sig-paren">(</span><em>chrom</em>, <em>position</em>, <em>ref</em>, <em>alt</em><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.get_variant_haplotypes_strict" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads phased genotypes at variant that matches specified chromosome, position, reference and alternate alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – chromosome name.</li>
<li><strong>position</strong> (<a class="reference external" href="https://docs.python.org/2/library/functions.html#long" title="(in Python v2.7)"><em>long</em></a>) – chromosomal position in base pairs.</li>
<li><strong>ref</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – reference allele.</li>
<li><strong>alt</strong> (<a class="reference external" href="https://docs.python.org/2/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) – alternate allele.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">phased genotypes at specified variant.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#csg.genetics.ld.pyld.Haplotypes" title="csg.genetics.ld.pyld.Haplotypes">Haplotypes</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="csg.genetics.ld.pyld.LD.release_vcfs">
<code class="descname">release_vcfs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#csg.genetics.ld.pyld.LD.release_vcfs" title="Permalink to this definition">¶</a></dt>
<dd><p>Closes all opened VCF files.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-csg.genetics.ld">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-csg.genetics.ld" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">csg.genetics.ld package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-csg.genetics.ld.pyld">csg.genetics.ld.pyld module</a></li>
<li><a class="reference internal" href="#module-csg.genetics.ld">Module contents</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Ryan Welch.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>