
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>csglib’s modules &#8212; csglib 0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="csglib-s-modules">
<h1>csglib’s modules<a class="headerlink" href="#csglib-s-modules" title="Permalink to this headline">¶</a></h1>
<div class="section" id="interval-tree">
<span id="id1"></span><h2>Interval tree<a class="headerlink" href="#interval-tree" title="Permalink to this headline">¶</a></h2>
<p>The <em>IntervalTree</em> module implements interval tree data structure for time efficient (log complexity) interval queries.
The implementation is based on red-black binary tree and supports the following queries:</p>
<ul class="simple">
<li>find all intervals overlapping given point;</li>
<li>find all intervals overlapping given interval;</li>
<li>find nearest interval to the left of the given point;</li>
<li>find nearest interval to the right of the given point;</li>
<li>find <em>K</em> nearest intervals to the left of the given point;</li>
<li>find <em>K</em> nearest intervals to the right of the given point;</li>
<li>get <em>K</em> first (leftmost) intervals;</li>
<li>get <em>K</em> last (rightmost) intervals;</li>
<li>traverse all intervals in ascending/descending order;</li>
<li>merge overlapping intervals;</li>
<li>construct complementary intervals (i.e. extracts all gaps between non-overlapping intervals).</li>
</ul>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">csg.intervaltree.IntervalTree</span> <span class="k">import</span> <span class="n">IntervalTree</span>

<span class="n">intervals</span> <span class="o">=</span> <span class="n">IntervalTree</span><span class="p">()</span> <span class="c1"># create new interval tree</span>

<span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">14</span><span class="p">)):</span> <span class="c1"># add intervals</span>
   <span class="n">intervals</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

<span class="nb">print</span> <span class="s1">&#39;Number of intervals:&#39;</span><span class="p">,</span> <span class="n">intervals</span><span class="o">.</span><span class="n">get_intervals_count</span><span class="p">()</span>

<span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">intervals</span><span class="o">.</span><span class="n">descending</span><span class="p">():</span> <span class="c1"># list intervals in descending order</span>
   <span class="nb">print</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span>

<span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">intervals</span><span class="o">.</span><span class="n">point_intersect</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span> <span class="c1"># find all intervals that intersect position 10</span>
   <span class="nb">print</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span>

<span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">intervals</span><span class="o">.</span><span class="n">interval_overlap</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">):</span> <span class="c1"># find all intervals that overlap interval [2, 5]</span>
   <span class="nb">print</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span>

<span class="n">merged_intervals</span> <span class="o">=</span> <span class="n">intervals</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span> <span class="c1"># merge all overlapping intervals</span>
<span class="nb">print</span> <span class="s1">&#39;Number of intervals after merging:&#39;</span><span class="p">,</span> <span class="n">merged_intervals</span><span class="o">.</span><span class="n">get_intervals_count</span><span class="p">()</span>
</pre></div>
</div>
<p>Please, refer to <a class="reference internal" href="csg.intervaltree.html"><span class="doc">API documentation</span></a> for further details.</p>
</div>
<div class="section" id="linkage-disequilibrium">
<span id="pyld"></span><h2>Linkage disequilibrium<a class="headerlink" href="#linkage-disequilibrium" title="Permalink to this headline">¶</a></h2>
<p>The <em>pyld</em> module implements computation of <em>r</em> coefficient of linkage disequilibrium (LD).
No special file format is required. Phased genotypes must be provided in VCF files compressed using bgzip and indexed using tabix.
Supported operations:</p>
<ul class="simple">
<li>compute pairwise LD between pair of SNPs from any given chromosomal region;</li>
<li>compute pairwise LD between a single SNP and all other SNPs from a given region;</li>
<li>compute pairwise LD between all SNPs from a given region;</li>
<li>compute alternate allele frequencies.</li>
</ul>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">csg.genetics.ld.pyld</span> <span class="k">import</span> <span class="n">LD</span>

<span class="n">ld</span> <span class="o">=</span> <span class="n">LD</span><span class="p">()</span>

<span class="n">ld</span><span class="o">.</span><span class="n">add_vcf</span><span class="p">(</span><span class="s1">&#39;genotypes.phased.vcf.gz&#39;</span><span class="p">)</span> <span class="c1"># open VCF with all chromosomes</span>
<span class="c1"># Alternatively, you may load VCF files by chromosome:</span>
<span class="c1"># ld.add_vcf(&#39;chr1.vcf.gz&#39;)</span>
<span class="c1"># ld.add_vcf(&#39;chr2.vcf.gz&#39;)</span>
<span class="c1"># ...</span>
<span class="c1"># ld.add_vcf(&#39;chr22.vcf.gz&#39;)</span>

<span class="n">haplotypes</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">get_region_haplotypes</span><span class="p">(</span><span class="s1">&#39;20&#39;</span><span class="p">,</span> <span class="mi">11650214</span><span class="p">,</span> <span class="mi">60759931</span><span class="p">)</span> <span class="c1"># read phased genotypes in 20:11650214-60759931</span>

<span class="n">freqs</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">compute_freq</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">)</span> <span class="c1"># compute alternative allele frequencies</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">haploypes</span><span class="o">.</span><span class="n">size</span><span class="p">):</span>
   <span class="nb">print</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">chrom</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">freqs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

<span class="n">r</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">compute_r_pairwise</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">)</span> <span class="c1"># compute LD between all variants in 20:11650214-60759931</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">size</span><span class="p">):</span>
   <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">size</span><span class="p">):</span>
      <span class="nb">print</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">chrom</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span>

<span class="n">haplotypes1</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">get_variant_haplotypes</span><span class="p">(</span><span class="s1">&#39;20&#39;</span><span class="p">,</span> <span class="mi">11650214</span><span class="p">)</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">compute_r_cross</span><span class="p">(</span><span class="n">haplotypes1</span><span class="p">,</span> <span class="n">haplotypes</span><span class="p">)</span> <span class="c1"># compute LD between variant 20:11650214 and all variants in 20:11650214-60759931</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">haplotypes1</span><span class="o">.</span><span class="n">size</span><span class="p">):</span>
   <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">size</span><span class="p">):</span>
      <span class="nb">print</span> <span class="n">haplotypes1</span><span class="o">.</span><span class="n">chrom</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">haplotypes1</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span>

<span class="n">haplotypes2</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">get_variant_haplotypes</span><span class="p">(</span><span class="s1">&#39;20&#39;</span><span class="p">,</span> <span class="mi">16655993</span><span class="p">)</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">ld</span><span class="o">.</span><span class="n">compute_r_cross</span><span class="p">(</span><span class="n">haplotypes1</span><span class="p">,</span> <span class="n">haplotypes2</span><span class="p">)</span> <span class="c1"># compute LD between vatiants 20:11650214 and 20:16655993</span>
<span class="nb">print</span> <span class="n">haplotypes1</span><span class="o">.</span><span class="n">chrom</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">haplotypes1</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">haplotypes2</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span>

<span class="n">ld</span><span class="o">.</span><span class="n">release_vcfs</span><span class="p">()</span> <span class="c1"># close VCF files</span>
</pre></div>
</div>
<p>Please, refer to <a class="reference internal" href="csg.genetics.ld.html"><span class="doc">API documentation</span></a> for further details.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">csglib’s modules</a><ul>
<li><a class="reference internal" href="#interval-tree">Interval tree</a></li>
<li><a class="reference internal" href="#linkage-disequilibrium">Linkage disequilibrium</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Ryan Welch.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
    </div>

    

    
  </body>
</html>